Updated docs for gambit 0.2.0 release (#88)
Certora/Documentation: docs/gambit/gambit.md

summarizing generated mutants in a human-readable way.

Running `gambit mutate` will invoke the solidity compiler via `solc`, so make
sure it is visible on your `PATH`. Alternatively, you can specify where Gambit can
find the Solidity compiler with the option `--solc path/to/solc`, or specify a
version of solc (e.g., solc8.12) with the option `--solc solc8.12`.

### Running  `gambit mutate` 

The `gambit mutate` command expects either a `--filename` argument or a `--json`
argument.  Using `--filename` allows you to specify a specific Solidity file to
mutate:

```bash
gambit mutate --filename file.sol
```

However, if you want to mutate multiple files or apply a more complex set of
parameters, we recommend using a configuration file via the `--json` option
instead:

```bash
gambit mutate --json gambit-conf.json
``` 

Run `gambit --help` for more information.

_**Note:** all relative paths specified in a JSON configuration file are interpreted
to be relative to the config file's parent directory._

In the following section we provide examples of how to run Gambit using both
`--filename` and `--json`. We provide more complete documentation in the

> [Configuration Files](#configuration-files) and [CLI-Options](#cli-options) sections below.

Reviewer: The correct syntax for myst references is
```
{ref}`ref-name`
```

where you've added a label to the referenced section, like:
```
(ref-name)=
## Referenced section
```

_**Note:** The mutated file must located within your current working directory or
one of its subdirectories. If you want to mutate code in an arbitrary directory,
use the `--sourceroot` option._

### Example 2: Mutating and Down Sampling

The above command produced 34 mutants which may be more than you need. Gambit
provides a way to randomly down sample the number of mutants with the
`--num-mutants` or `-n` option:

```bash
gambit mutate -f benchmarks/BinaryOpMutation/BinaryOpMutation.sol -n 3
```
<pre>
Generated 3 mutants in 0.15 seconds
</pre>

### Example 3: Viewing Gambit Results
_**Note:** this example assumes you've just completed Example 2_

Gambit outputs all of its results in `gambit_out`:

```bash
tree -L 2 gambit_out
```
<!-- Code output: using `pre` to avoid the Copy To Clipboard feature -->
<pre>
gambit_out
├── gambit_results.json
├── input_json
│   ├── BinaryOpMutation.sol_json.ast
│   └── BinaryOpMutation.sol_json.ast.json
├── mutants
│   ├── 1
│   ├── 2
│   └── 3
└── mutants.log
</pre>

See the [Results Directory](#results-directory) section for a detailed
explanation of this layout. The `gambit summary` command
pretty prints each mutant for easy inspection:

> ![The output of `gambit summary`](doc/gambit-summary.png)

Reviewer: You need to add the image

# Mutant Generation

Gambit is a state-of-the-art mutation system for Solidity.
By applying predefined syntax transformations called _mutation operators_ (for
  example, `a + b` -> `a - b`) to a Solidity program's source code, Gambit
  generates variants of the program called _mutants_.
Mutants can be used to evaluate test suites or specs used for formal
  verification: each mutant represents a potential bug in the program, and
  stronger test suites and specifications should detect more mutants.

## Requirements

1. Gambit is written in Rust. You'll need to [install Rust and
   Cargo](https://www.rust-lang.org/tools/install) to build Gambit.

> 2. Gambit uses the solc, the Solidity compiler, to generate mutants. You'll need

Reviewer: ```suggestion
2. Gambit uses `solc`, the Solidity compiler, to generate mutants. You'll need
```

>    to have solc binary that is compatible with the project you are mutating (see

Reviewer: ```suggestion
   to have a `solc` binary that is compatible with the project you are mutating (see
```

   the `--solc` option in `gambit mutate --help`)

## Installation

You can download prebuilt Gambit binaries for Mac and Linux from our

> [releases](https://github.com/Certora/gambit/releases) page.

Reviewer: not a big deal, but I prefer the footnote-style links
```suggestion
[releases][gambit-releases] page.

[releases]: https://github.com/Certora/gambit/releases
```

> To build Gambit from source, clone this repository and run

Reviewer: ```suggestion
To build Gambit from source, clone [the gambit repository][gambit-repo] and run
```

Reviewer: and add the link

```
cargo install --path .
```

from this repository's root. This will build Gambit and install it to a globally visible
location on your `PATH`.

You can also build gambit with `cargo build --release` from the root of this
repository.  This will create a `gambit` binary in `gambit/target/release/`
which you can manually place on your path or invoke directly (e.g., by calling
`path/to/gambit/target/release/gambit`).

## Usage

Gambit has two main commands: `mutate` and `summary`. `gambit mutate` is
responsible for mutating code, and `gambit summary` is a convenience command for
summarizing generated mutants in a human-readable way.

> Running `gambit mutate` will invoke the solidity compiler via `solc`, so make

Reviewer: ```suggestion
Running `gambit mutate` will invoke the Solidity compiler using `solc`, so make
```

sure it is visible on your `PATH`. Alternatively, you can specify where Gambit can
find the Solidity compiler with the option `--solc path/to/solc`, or specify a

> version of solc (e.g., solc8.12) with the option `--solc solc8.12`.

Reviewer: ```suggestion
`solc` binary (e.g., `solc8.12`) with the option `--solc solc8.12`.
```

### Running  `gambit mutate` 

The `gambit mutate` command expects either a `--filename` argument or a `--json`
argument.  Using `--filename` allows you to specify a specific Solidity file to
mutate:

```bash
gambit mutate --filename file.sol
```

However, if you want to mutate multiple files or apply a more complex set of
parameters, we recommend using a configuration file via the `--json` option
instead:

```bash
gambit mutate --json gambit-conf.json
``` 

Run `gambit --help` for more information.

_**Note:** all relative paths specified in a JSON configuration file are interpreted
to be relative to the config file's parent directory._

In the following section we provide examples of how to run Gambit using both
`--filename` and `--json`. We provide more complete documentation in the
{ref}`configuration-files` and {ref}`cli-options` sections below.

## Examples

> Unless otherwise noted, examples use code from `benchmarks/` and are run from

Reviewer: Make this a link to the folder on github

> the root of this repository.

Reviewer: change "this repository" to "the gambit repository" and make it a link

> ### Example 1: Mutating a Single File

Reviewer: ```suggestion
### Example 1: Mutating a single file
```

To mutate a single file, use the `--filename` option (or `-f`), followed by the
file to mutate.

```bash
gambit mutate -f benchmarks/BinaryOpMutation/BinaryOpMutation.sol
```

> <!-- Code output: using `pre` to avoid the Copy To Clipboard feature -->

Reviewer: any particular reason?  the formatting looks funny to me

Reviewer: I would surround the example with "For example, [command] will output [output]"

<pre>
Generated 34 mutants in 0.69 seconds
</pre>

> _**Note:** The mutated file must located within your current working directory or

Reviewer: ```suggestion
_**Note:** The mutated file must be located within your current working directory or
```

> _**Note:** The mutated file must located within your current working directory or

Reviewer: Prefer
````
```{note}
<note contents>
```
````
over hand-formatted notes

use the `--sourceroot` option._

### Example 2: Mutating and Down Sampling

The above command produced 34 mutants which may be more than you need. Gambit
provides a way to randomly down sample the number of mutants with the
`--num-mutants` or `-n` option:

```bash
gambit mutate -f benchmarks/BinaryOpMutation/BinaryOpMutation.sol -n 3
```
<pre>
Generated 3 mutants in 0.15 seconds
</pre>

### Example 3: Viewing Gambit Results
_**Note:** this example assumes you've just completed Example 2_

Gambit outputs all of its results in `gambit_out`:

```bash
tree -L 2 gambit_out
```
<!-- Code output: using `pre` to avoid the Copy To Clipboard feature -->
<pre>
gambit_out
├── gambit_results.json
├── input_json
│   ├── BinaryOpMutation.sol_json.ast
│   └── BinaryOpMutation.sol_json.ast.json
├── mutants
│   ├── 1
│   ├── 2
│   └── 3
└── mutants.log
</pre>

See the {ref}`results-directory` section for a detailed
explanation of this layout. The `gambit summary` command
pretty prints each mutant for easy inspection:

![The output of `gambit summary`](gambit-summary.png)

By default `gambit summary` prints info on all mutants. If you are interested in

> particular mutants you can specify a subset of mutant ids with the `--mids` flag.

Reviewer: or maybe just "mutants" instead of "mutant IDs", since you are printing the mutants, after all.

For instance, `gambit summary --mids 3 4 5`  will only print info for mutant ids
3 through 5.

### Example 4: Specifying solc Pass-through Arguments
Solc may need some extra information to successfully run on a file or a project.
Gambit enables this with _pass-through arguments_ that, as the name suggests,

> are passed directly through to the solc compiler.

Reviewer: `solc`

For projects that have complex dependencies and imports, you may need to:

> * **Specify base-paths**: To specify the Solidity [--base-path][basepath]

Reviewer: ```suggestion
* **Specify base paths**: To specify the Solidity [`--base-path`][basepath]
```

```bash
gambit mutate --json benchmarks/config-jsons/test1.json
```

The configuration file is a JSON file containing the command line arguments for
`gambit` and additional configuration options:

```json
{
    "filename": "../10Power/TenPower.sol",
    "sourceroot": "..",
    "solc-remappings": [
        "@openzeppelin=node_modules/@openzeppelin"
    ],
}
```

In addition to specifying the command line arguments, you can list the specific
mutants that you want to apply, the specific functions you wish to mutate, and
more.  See the [`benchmark/config-jsons` directory][config-examples] for
examples.

_**Note:** Any paths provided by the configuration file are resolved relative to
the configuration file's parent directory._

(configuration-files)=
## Configuration Files
Configuration files allow you to save complex configurations and perform
multiple mutations at once. Gambit uses a simple JSON object format to store
mutation options, where each `--option VALUE` specified on the CLI is
represented as a `"option": VALUE` key/value pair in the JSON object.  Boolean
`--flag`s are enabled by storing them as true: `"flag": true`. For instance,

> `--no-overwrite` would be written as `"no-overwrite": true"`.

Reviewer: ```suggestion
`--no-overwrite` would be written as `"no-overwrite": true`.
```
